wordpress-mobile/WordPress-Android
# Improve deeplink tracking

Fixes #14647 

I'm sorry for the size of this PR. It unexpectedly grew over the time with the changing requirements. What it contains is:
- strip the URL tracked with each deep link click of any data containing any identifiers
- add `source` to each event - tracking where the link came from - email, banner or link
- add `source_info` to each event - additional information - in our case only `login_reason` from the `create_site` links

In order to achieve this I've extracted a `DeepLinkHandler` interface and create a provider for all the handlers. I think the resulting structure is neat and simplifies the `ViewModel` a lot. Let me know what you think! I've tried to cover everything with unit tests.

For the `read` URLs from the `ReaderLinkHandler` that are now redirected to the `ReaderPostPagerActivity` - these are now tracked twice. I think that's the cleanest approach at the moment. Both of these have a different `action` so it should be easy to differentiate between these two events. Let me know what you think about this approach!

To test:
- Try various deeplinks and app links (`wordpress.com`, `public-api.wordpress.com` tracking links, `wordpress://...` app links) 
- Check the `deep_linked` event in Logcat
- Notice it has the following properties:
  - intent_host - it was there before and is the host of the original link
  - source
  - url (which now contains the stripped URL)
  - source_info (only there for the `Create a site` links if `login_reason` is present in the original tracking URL)

## Regression Notes
1. Potential unintended areas of impact
- All the deeplinks handling

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- I've tested all of the deeplinks I could think of and covered everything with unit tests

3. What automated tests I added (or what prevented me from doing so)
- I've added a ton of unit tests

PR submission checklist:
